core/src/main/java/org/infinispan/conflict/impl/StateReceiverImpl.java

```diff
@@ -82,45 +67,75 @@ public void start() {
-         synchronized (lock) {
-            if (trace) log.tracef("Received DataRehashedEvent %s", dataRehashedEvent);
-
-            boolean newSegmentOwners = newSegmentOwners(dataRehashedEvent);
+         log.debugf("Cancelling all segment requests as rehash has started %s", dataRehashedEvent);
 
-            if (minTopologyId < 0 || newSegmentOwners)
-               minTopologyId = dataRehashedEvent.getNewTopologyId();
-
-            if (newSegmentOwners && !segmentRequestFuture.isDone())
-               cancelAllSegmentRequests(new CacheException("Cancelling replica request as the owners of the requested " +
-                     "segment have changed."));
-         }
+         for (SegmentRequest request : requestMap.values())
+            request.cancel(new CacheException("Cancelling replica request as the owners of the requested " +
+               "segment have changed."));
       }
    }
 
    @Override
    public CompletableFuture<List<Map<Address, InternalCacheEntry<K, V>>>> getAllReplicasForSegment(int segmentId, LocalizedCacheTopology topology) {
-      synchronized (lock) {
-         // TODO improve handling of concurrent invocations
-         if (!segmentRequestFuture.isDone()) {
-            if (trace) log.tracef("Concurrent invocations of getAllReplicasForSegment detected %s", segmentRequestFuture);
-            throw new IllegalStateException("It is not possible to request multiple segments of a cache simultaneously.");
-         }
+      return requestMap.computeIfAbsent(segmentId, id -> new SegmentRequest(id, topology)).requestState();
+   }
+
+   @Override
+   public void receiveState(Address sender, int topologyId, Collection<StateChunk> stateChunks) {
+      if (stateChunks.isEmpty()) {
+         if (trace)
+            log.tracef("Ignoring received state for cache %s from %s because stateChunks are empty", cacheName, sender);
+         return;
+      }
+
+      int segmentId = stateChunks.iterator().next().getSegmentId();
+      SegmentRequest request = requestMap.get(segmentId);
+      if (request == null) {
+         if (trace) log.tracef("Ignoring received state for cache %s because the associated request was completed or cancelled %s", cacheName);
+         return;
+      }
+      request.receiveState(sender, topologyId, stateChunks);
+   }
+
+   Map<K, Map<Address, InternalCacheEntry<K, V>>> getKeyReplicaMap(int segmentId) {
+      return requestMap.get(segmentId).keyReplicaMap;
+   }
+
+   Map<Address, InboundTransferTask> getTransferTaskMap(int segmentId) {
+      return requestMap.get(segmentId).transferTaskMap;
+   }
+
+   InboundTransferTask createTransferTask(int segmentId, Address source, CacheTopology topology) {
+      return new InboundTransferTask(Collections.singleton(segmentId), source, topology.getTopologyId(),
+            rpcManager, commandsFactory, transferTimeout, cacheName, false);
+   }
+
+   class SegmentRequest {
```

Comment:
I refactored the code to allow for concurrent segment requests, as previously the `StateReceiverImpl` only allowed a single segment request to be executing at one time.